pkg/client/conditions.go

```diff
@@ -22,13 +22,19 @@ import (
 )
 
 // ControllerHasDesiredReplicas returns a condition that will be true iff the desired replica count
-// for a controller's ReplicaSelector equals the Replicas count.
+// for a controller's ReplicaSelector equals the Replicas count, and the sequence number of the
+// controller's status is at least as high as the sequence number of the incoming controller.
 func ControllerHasDesiredReplicas(c Interface, controller *api.ReplicationController) wait.ConditionFunc {
+
+	// If we're given a controller where the status lags the spec, it either means that the controller is stale,
+	// or that the rc manager hasn't noticed the update yet. Polling status.Replicas is not safe in the latter case.
+	desiredSequenceNumber := controller.Spec.SequenceNumber
+
 	return func() (bool, error) {
 		ctrl, err := c.ReplicationControllers(controller.Namespace).Get(controller.Name)
 		if err != nil {
 			return false, err
 		}
-		return ctrl.Status.Replicas == ctrl.Spec.Replicas, nil
+		return ctrl.Status.SequenceNumber >= desiredSequenceNumber && ctrl.Status.Replicas == ctrl.Spec.Replicas, nil
```

Comment:
This should use ctrl.Spec.SequenceNumber. Otherwise, it's wrong in the case that spec.Replicas is updated while this is polling.